File: lista_encadeada/lista_encadeada.py
from no import No

class ListaEncadeada:
  def __init__(self):
    self.primeiro = None

  def inserir(self, elemento):
    if self.consultar(elemento) is not None:
      return False

    no = No(elemento)
    no.set_proximo(self.primeiro)
    self.primeiro = no
    return True

  def consultar(self, elemento):
    atual = self.primeiro
    while atual is not None:
      if atual.get_elemento() == elemento:
        return True
      atual = atual.get_proximo()

    return False

  def remover(self, elemento):
    atual = self.primeiro
    anterior = None

    while atual is not None:
      if atual.get_elemento() == elemento:
        if anterior is None:
          self.primeiro = atual.get_proximo()
        else:
          anterior.set_proximo(atual.get_proximo())
        return True

      anterior = atual
      atual = atual.get_proximo()

    return False
    
lista = ListaEncadeada()
lista.adicionar(1)
lista.adicionar(2)
lista.adicionar(3)
print(lista)

File: estrutura-de-dados-em-java/src/lista/ListaEncadeada.java
package lista;

public class ListaEncadeada<T> {
  private No<T> inicio;
  private No<T> ultimo;
  private int tamanho = 0;

  public void adicionar(T elemento) {
    No<T> no = new No<T>(elemento);

    if (this.tamanho == 0) {
      this.inicio = no;
    } else {
      this.ultimo.setProximo(no);
    }
    
    this.ultimo = no;
    this.tamanho++;
  }

  public int getTamanho() {                                                                                                                                                                                                         
    return this.tamanho;
  }

  @Override
  public String toString() {
    if (this.tamanho == 0) {
      return "[]";
    }

    StringBuilder builder = new StringBuilder("[");
    // builder.append("ListaEncadeada [inicio=").append(inicio).append("]");

    No<T> atual = this.inicio;
    int i = 0;
    while (i < this.tamanho - 1) {
      builder.append(atual.getElemento()).append(",");
      atual = atual.getProximo();
      i++;
    }
    builder.append(atual.getElemento()).append("]");
    return builder.toString();
  }
}